### PR TITLE
[BUGFIX] Allow "Content-Type" in content

### LINKED ISSUE #957: False positives for content-type meta element presence check
I've recently updated Emogrifier and when moving from the old style syntax to the new, I receive this error:
`PHP Fatal error:  Uncaught TypeError: Return value of Pelago\Emogrifier\CssInliner::getHeadElement() must be an instance of DOMElement, null returned in /var/www/vhosts/test.system/httpdocs/vendor/pelago/emogrifier/src/CssInliner.php:1293`

Is there a reason there must always be a `<head>` element?
https://github.com/MyIntervals/emogrifier/blob/b43b650880d189b0ada61d95d0729c7424b1752d/src/CssInliner.php#L1291

I'm just inlining CSS in a simple email template that doesn't use a `<head>` element for anything.
I've added a blank one for now but strikes me as a little arbitrary.

Thanks!

### PR DESCRIPTION
**Pull request by JakeQZ**
Use a regular expression to determine whether the provided HTML contains a
`Content-Type` `meta` element instead of just testing for the string
"Content-Type".  Also allow for case insensitivity.

A valid `Content-Type` `meta` element which is invalidly placed (i.e. within the
`body`) will still be determined to be valid.  This will be addressed separately
for #923.

Add the package `rawr/cross-data-providers` to the development dependencies.
This is useful to generate a matrix of combinations for test data providers.
There are a few existing instances where this could be used to replace code
effectively doing that, but which are not addressed in this commit.

Rename affected data provider methods from `xyzDataProvider` to `provideXyx`,
which is more succinct and a more commonly used nomenclature on other projects.

Fixes #957.

### CODE REVIEW

File: src/HtmlProcessor/AbstractHtmlProcessor.php [hunk 1]
```diff
@@ -271,7 +271,8 @@ private function normalizeDocumentType(string $html): string
      */
     private function addContentTypeMetaTag(string $html): string
     {
-        $hasContentTypeMetaTag = \stripos($html, 'Content-Type') !== false;
+        $hasContentTypeMetaTag
+            = \preg_match('%<meta(?=\\s)[^>]*\\shttp-equiv=(["\']?+)Content-Type\\g{-1}[\\s/>]%i', $html) === 1;
```
Comment (oliverklee) [inline src/HtmlProcessor/AbstractHtmlProcessor.php:275]: With the `=== 1`,  we've switched the meaning from "at least one content-type meta tag" to "exactly one content-type meta tag".

Even if multiple occurrences do not make much sense, I'd still prefer us to recognize this as "there is one" by keeping using `!== false`, and also cover this edge case with a test.

Comment (oliverklee) [reply src/HtmlProcessor/AbstractHtmlProcessor.php:275]: I don't quite understand the `(?=\\s)` part yet. What does it do?

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: src/HtmlProcessor/AbstractHtmlProcessor.php (hunk 1 above)
Comment (JakeQZ) [reply src/HtmlProcessor/AbstractHtmlProcessor.php:275]: > 
> 
> With the `=== 1`, we've switched the meaning from "at least one content-type meta tag" to "exactly one content-type meta tag".
> 
> Even if multiple occurrences do not make much sense, I'd still prefer us to recognize this as "there is one" by keeping using `!== false`, and also cover this edge case with a test.

The return value does appear to indicate the number of matches (`0` or `1` because it is limited to one match) with `false` on error.  So to be more explicit we'd probably need something like

```php
$contentTypeMetaTagMatchCount = \preg_match(...);
$hasContentTypeMetaTag = $contentTypeMetaTagMatchCount !== false && $contentTypeMetaTagMatchCount > 0;
```

Comment (JakeQZ) [reply src/HtmlProcessor/AbstractHtmlProcessor.php:275]: > 
> 
> I don't quite understand the `(?=\\s)` part yet. What does it do?

It asserts that at that position the next character is a whitespace but without 'consuming' it ('lookahead').  If it was 'consumed', `\\shttp-equiv` would not match if there was nothing else beetween `meta` and `http-equiv` because the whitespace to be mached there would have already been consumed.  Without asserting `meta` is followed by whitespace, other tag names beginning with `meta` (hitherto as yet not defined) would also be matched.

Comment (oliverklee) [reply src/HtmlProcessor/AbstractHtmlProcessor.php:275]: > So to be more explicit we'd probably need something like
> 
> ```
> $contentTypeMetaTagMatchCount = \preg_match(...);
> $hasContentTypeMetaTag = $contentTypeMetaTagMatchCount !== false && $contentTypeMetaTagMatchCount > 0;
> ```

Or even more explicit:

```php
$hasContentTypeMetaTag = \is_int($contentTypeMetaTagMatchCount) false && $contentTypeMetaTagMatchCount > 0;
```

Comment (oliverklee) [reply src/HtmlProcessor/AbstractHtmlProcessor.php:275]: Thanks for the explanation!

**Comment by JakeQZ** [comment]
I've changed the test for the `preg_match` return value as suggested, and also added tests to confirm the regex does not match unexpected longer keywords which contain the expected keyword as a substring.

**Review by oliverklee — APPROVED** [review]
LGTM.
